### PR TITLE
raftstore: cancel gen snap task if a peer becomes unreachable

### LINKED ISSUE #17191: tikv scaling in blocked due to one region can not transfer
### What version of TiKV are you using?
[2024/06/25 11:43:54.651 +08:00] [INFO] [cmd.go:151] ["Start remote command"] [cmd="/tiup/deploy/tikv-*/bin/tikv-server -V"] [nodename=tikv2]
TiKV 
Release Version:   8.2.0-alpha
Edition:           Community
Git Commit Hash:   87f66434fb7fa2395cb5ebfe45b3f66e2527c451
Git Commit Branch: HEAD
UTC Build Time:    2024-06-24 06:44:26
Rust Version:      rustc 1.77.0-nightly

[2024/06/25 11:43:55.616 +08:00] [INFO] [cmd.go:151] ["Start remote command"] [cmd="/tiup/deploy/pd-*/bin/pd-server -V"] [nodename=pd2]
Release Version: v8.2.0-alpha-97-gdebb5fe
Edition: Community
Git Commit Hash: debb5febb83e032da5a0c0162b1bff656a8fe931
Git Branch: HEAD
UTC Build Time:  2024-06-24 03:24:39

### What operating system and CPU are you using?
8c/16g

### Steps to reproduce
1、scale out tikv from 3 to 6
2、wait for scale out to complete
3、scale in tikv from 6 to 3

### What did you expect?
scale in tikv success

### What did happened?
the status of one tikv is always offline due to one region can not transfer
img_v3_02c7_125604a5-1a00-4992-aabd-759b2523bf6g
image

### LINKED ISSUE #17213: Raft snapshot blocked when peer is removed right after requesting snapshot
## Bug Report

Raft snapshots get blocked when a peer is removed immediately after requesting a snapshot. This occurs because the leader must wait for `MsgSnapGenPrecheckResponse` before generating snapshots. As a result, not only is the snapshot blocked, but subsequent conf changes (such as adding a node) are also blocked.

To resolve this issue, TiKV needs to cancel the snapshot generation task if a peer becomes unreachable.

https://github.com/tikv/tikv/blob/5229f839c48d3b3f3f1218e7a465e13f37e864e8/components/raftstore/src/store/peer.rs#L2658-L2676

Cc #17019 

### What version of TiKV are you using?

Nightly, 2024-06-26

### Steps to reproduce

```rust
#[test_case(test_raftstore::new_node_cluster)]
fn test_remove_peer_before_receiving_snapshot() {
    let mut cluster = new_cluster(0, 3);
    let pd_client = cluster.pd_client.clone();
    pd_client.disable_default_operator();

    // Use run conf change to make new node be initialized with term 0.
    let r = cluster.run_conf_change();

    // Add peer(3, 3) to store 3
    cluster.must_put(b"k0", b"v0");
    pd_client.must_add_peer(r, new_peer(3, 3));
    must_get_equal(&cluster.get_engine(3), b"k0", b"v0");

    // Add peer(2, 2) to store 2
    //
    // Drop MsgSnapGenPrecheckRequest to region 1 on store 2 before adding peer.
    let (send_tx, send_rx) = mpsc::sync_channel(1);
    cluster.add_send_filter(CloneFilterFactory(
        RegionPacketFilter::new(r, 2)
            .msg_type(MessageType::MsgSnapshot)
            .drop_extra_message(ExtraMessageType::MsgSnapGenPrecheckRequest)
            .direction(Direction::Recv)
            .set_msg_callback(Arc::new(move |m: &RaftMessage| {
                if m.get_extra_msg().get_type() == ExtraMessageType::MsgSnapGenPrecheckRequest {
                    let _ = send_tx.send(());
                }
            })),
    ));
    // Make sure add peer conf change is applied on leader side.
    pd_client.must_add_peer(r, new_peer(2, 2));

    // Make sure peer 2 has requested snapshot.
    send_rx.recv_timeout(Duration::from_secs(5)).unwrap();

    // Remove peer(2, 2) from store 2
    pd_client.must_remove_peer(r, new_peer(2, 2));

    // Clear filters and add peer(2, 4) to store 2.
    cluster.clear_send_filters();
    pd_client.must_add_peer(r, new_peer(2, 4));

    // Transfer leader to peer 4 to make sure it has applied a snapshot.
    cluster.must_transfer_leader(1, new_peer(2, 4));
    cluster.must_put(b"k1", b"v1");
    must_get_equal(&cluster.get_engine(2), b"k1", b"v1");
}
```

### What did you expect?

Dot not block snapshot.

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: close #17213 close #17191

What's Changed:

```commit-message
After #17019, raft snapshots get blocked when a peer is removed
immediately after requesting a snapshot. This occurs because the leader
must wait for `MsgSnapGenPrecheckResponse` before generating snapshots.
As a result, not only is the snapshot blocked, but subsequent conf
changes (such as adding a node) are also blocked.

To resolve this issue, TiKV needs to cancel the snapshot generation
task if a peer becomes unreachable.
```

### Related changes

- Need to cherry-pick to the release branch

### Check List

Tests 

- Integration test

### Release note

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/peer.rs [hunk 1]
```diff
@@ -2655,26 +2651,51 @@ where
...
+    pub fn maybe_cancel_gen_snap_task(&self, unreachable_store_id: Option<u64>) {
+        let to_peer = match self.get_store().get_gen_snap_task().as_ref() {
+            Some(task) => task.to_peer.clone(),
+            None => return,
+        };
+        let cancel_by_unreachable_store =
+            unreachable_store_id.map_or(false, |s| to_peer.get_store_id() == s);
+        let to_peer = to_peer.clone();
+        let cancel_by_peer_not_found = find_peer_by_id(self.region(), to_peer.get_id()).is_none();
+        if cancel_by_unreachable_store || cancel_by_peer_not_found {
```
Comment (hbisheng) [inline components/raftstore/src/store/peer.rs:2668]: The two canceling scenarios here look reasonable to me. I might be overthinking but if we miss some other scenarios that should be considered then we might run into the same issue again.

Not strictly related to this PR, do we consider having an online config option to turn off the precheck mechanism? That might be an overkill though.

**Review by hbisheng — APPROVED** [review]
LGTM

**Review by glorv — APPROVED** [review]

**Pull request merged**